File: source/cpp/2306. Naming a Company.cpp
#include <string>
#include <unordered_set>
#include <vector>


class Solution {
public:
    long long distinctNames(std::vector<std::string>& ideas) {
        // Group idea by their initials.
        std::unordered_set<std::string> initialGroup[26];
        for (auto& idea : ideas)
            initialGroup[idea[0] - 'a'].insert(idea.substr(1));
        
        // Calculate number of valid names from every pair of groups.
        auto answer = 0LL;
        for (int i = 0; i < 25; ++i) {
            for (int j = i + 1; j < 26; ++j) {
                // Get the number of mutual suffixes.
                int numOfMutual = 0;
                for (auto& ideaA : initialGroup[i]) {
                    if (initialGroup[j].count(ideaA))
                        numOfMutual++;
                }

                // Valid names are only from distinct suffixes in both groups.
                // Since we can swap a with b and swap b with a to create two valid names, multiple answer by 2.
                answer += (initialGroup[i].size() - numOfMutual) * (initialGroup[j].size() - numOfMutual) * 2;
            }
        }

        return answer;
    }
};

File: source/python/2306. Naming a Company.py
class Solution:
    def distinctNames(self, ideas: list[str]) -> int:
        # Group the ideas by their first letter
        initial_groups: list[set[str]] = [set() for _ in range(26)]
        for idea in ideas:
            initial_groups[ord(idea[0]) - ord("a")].add(idea[1:])

        # Remove empty groups
        initial_groups = [group for group in initial_groups if group]

        answer = 0
        # Calculate the number of valid names from every pair of groups
        for i in range(len(initial_groups) - 1):
            for j in range(i + 1, len(initial_groups)):
                num_of_mutual = len(initial_groups[i] & initial_groups[j])

                # Valid names are only from distinct suffixes from both groups
                # Since we can a, b and b, a, we need to multiply by 2
                answer += (len(initial_groups[i]) - num_of_mutual) * \
                    (len(initial_groups[j]) - num_of_mutual) * 2

        return answer


def test():
    s = Solution()

    assert s.distinctNames(["coffee", "donuts", "time", "toffee"]) == 6
    assert s.distinctNames(["lack", "back"]) == 0

    print("All tests passed!")


if __name__ == "__main__":
    test()
